File: php/bob/Bob.php
<?php declare(strict_types=1);

class Bob
{
    public function respondTo(string $str): string
    {
        $str = trim($str);
        
        return match(true) {
            mb_substr($str, -1) === "?" => 
                match(true) {
                    !preg_match('/[a-z]/', $str) && preg_match('/[A-Z]+/', $str) => "Calm down, I know what I'm doing!",
                    default => "Sure."
                },
            !preg_match('/[a-z]/', $str) && preg_match('/[A-Z]+/', $str) => 'Whoa, chill out!',
            strlen($str) === 0 => 'Fine. Be that way!',
            default => "Whatever."
        };
    }
}

File: ruby/bob/bob.rb
class Bob
  def self.hey(remark)
    remark.strip!
    remark = remark.gsub(/\n/, " ")

    if remark.length == 0
      return "Fine. Be that way!"
    end

    if (/[A-Z]+\?$/ =~ remark) != nil && remark.upcase == remark
      return "Calm down, I know what I'm doing!"
    end

    if (/^.*\?$/ =~ remark) != nil
      return "Sure."
    end

    if (/[A-Z]+/ =~ remark) != nil && remark.upcase == remark
      return "Whoa, chill out!"
    end

    "Whatever."
  end
end
